### ISSUE TITLE
CNTK changing model topology and retrain

### ISSUE THREAD

**Issue by PatrickBue** [Microsoft/CNTK#430]
Hi,

I was wondering if CNTK supports fine-tuning an already trained AlexNet, and if there is any documentation/examples for that.
To be more specific, I trained the net using CNTK, now would like to remove the last 1000-way softmax layer, instead add a 2-way softmax, and then run a few training iterations. Ideally, I would want to not retrain the whole net, but just a few of the last layers.

Thanks!
Patrick

**Comment by Alexey-Kamenev** [comment]
Hi Patrick,
You can modify the network topology using MEL (Model Editing Language), please see section 4.2 in CNTK book. As for using the pre-trained net as feature extractor, so far I can only think of modifying relevant LearnableParameters and setting learningRateMultiplier parameter to 0 to prevent training.

**Comment by PatrickBue** [comment]
Thanks! I am going to look into this now..

**Comment by PatrickBue** [comment]
Below is my solution for loading a trained network, and replacing the last layer so that the network can be used for a different prediction problem. Not sure if there's an easier way to do this, but it seems to work.

I used the MNIST example scripts/data to get started. Each MNIST image is assigned one of 10 labels, according to the number it shows. As new prediction problem, I used the same images but changed the labels to be 0 if the digit is smaller than 5, and 1 otherwise. I verified that, by modifying the original MNIST network, only 1-2 training iterations were required to get good results for the binary dataset. In comparison, when training a network from scratch, at least 5-10 iterations were required.

Steps:
-  Trained an MNIST model using the provided example scripts (02_Convolution.cntk).
-  Modified the final model using the “edit” command with the MEL code below. This replaces the final 10-way softmax layer with a 2-way softmax.
-  Made a copy of 02_Convolution.cntk and modified the file slightly to read in my binary dataset. Verified that the script runs by training a new (binary) model from scratch.
-  Lastly, just like resuming training from a certain epoch, renamed the edited model (to e.g. 02_ConvolutionBINARY.0), dropped it into the model output directory, and ran the training scripts.  



#### MY MEL SCRIPT ####

#load pre-trained model
model1 = LoadModel("$RootDir$/02_Convolution")
SetDefaultModel(model1)
#DumpModel(model1, "$RootDir$/dump_originalModel.txt", includeData =true)

#add new final layer
inDim=128
labelDimNew=2
newL = DNNLayer(inDim, labelDimNew, h1.y, 1)
labelsNew = Input(labelDimNew, tag="label")
SetInput(ce, 0, labelsNew)
SetInput(ce, 1, newL.z)
SetInput(err, 0, labelsNew)
SetInput(err, 1, newL.z)
SetProperty(newL.z, "output", "true")

#remove old final layer (note:make sure these deletes happen in reverse order from the original definition)
DeleteNode(ol.z) 
DeleteNode(ol.t)
DeleteNode(ol.b)
DeleteNode(ol.W)
DeleteNode(labels)

#rename nodes to have same name as before (so that the same .cntk training script can be used)
Rename(labelsNew, labels)
Rename(newL.*, ol.*)

#save and dump model
SaveModel(model1, "$RootDir$/editedModel", format="cntk")
#DumpModel(model1, "$RootDir$/dump_editedModel.txt", includeData =true)

**Comment by Alexey-Kamenev** [comment]
This is great, thanks a lot. I think we should add this to our samples.

**Comment by PatrickBue** [comment]
No problem! I did not look into excluding a few of the early layers from training (see your comment regarding the learningRateMultiplier). That might be interesting as well.
